// mindustry-schematic-parser/dist/mindustry/block/blocks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blocks = void 0;
var tslib_1 = require("tslib");
var campaign = tslib_1.__importStar(require("./campaign"));
var crafting = tslib_1.__importStar(require("./crafting"));
var defense = tslib_1.__importStar(require("./defense"));
var distribution = tslib_1.__importStar(require("./distribution"));
var environment = tslib_1.__importStar(require("./environment"));
var experimental = tslib_1.__importStar(require("./experimental"));
var liquid = tslib_1.__importStar(require("./liquid"));
var logic = tslib_1.__importStar(require("./logic"));
var power = tslib_1.__importStar(require("./power"));
var production = tslib_1.__importStar(require("./production"));
var sandbox = tslib_1.__importStar(require("./sandbox"));
var storage = tslib_1.__importStar(require("./storage"));
var turrets = tslib_1.__importStar(require("./turrets"));
var units = tslib_1.__importStar(require("./units"));
var block_1 = require("./block");
function registerBlocks(domain) {
    for (var k in domain) {
        var key = k;
        var Class = domain[key];
        if (Class) {
            var block = new Class();
            block_1.Block.codes.set(block.name, block);
        }
    }
}
registerBlocks(campaign);
registerBlocks(crafting);
registerBlocks(tslib_1.__assign(tslib_1.__assign({}, defense), { Wall: undefined }));
registerBlocks(distribution);
registerBlocks(environment);
registerBlocks(experimental);
registerBlocks(liquid);
registerBlocks(logic);
registerBlocks(tslib_1.__assign(tslib_1.__assign({}, power), { PowerGenerator: undefined }));
registerBlocks(production);
registerBlocks(tslib_1.__assign(tslib_1.__assign({}, sandbox), { LightBlock: undefined }));
registerBlocks(storage);
registerBlocks(turrets);
registerBlocks(units);
exports.Blocks = {
    campaign: campaign,
    crafting: crafting,
    defense: defense,
    distribution: distribution,
    environment: environment,
    experimental: experimental,
    liquid: liquid,
    logic: logic,
    power: power,
    production: production,
    sandbox: sandbox,
    storage: storage,
    turrets: turrets,
    units: units,
};
