// mindustry-schematic-parser/dist/mindustry/block/campaign.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterplanetaryAccelerator = exports.LaunchPadLarge = exports.LaunchPad = void 0;
var tslib_1 = require("tslib");
var util_1 = require("../../util");
var block_1 = require("./block");
var category = 'campaign';
var Pad = /** @class */ (function (_super) {
    tslib_1.__extends(Pad, _super);
    function Pad() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Pad.prototype.draw = function (tile, canvas) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.render({
                            tile: tile,
                            canvas: canvas,
                            category: category,
                            layers: [this.name],
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return Pad;
}(block_1.Block));
var LaunchPad = /** @class */ (function (_super) {
    tslib_1.__extends(LaunchPad, _super);
    function LaunchPad() {
        return _super.call(this, {
            name: 'launch-pad',
            requirements: { copper: 350, silicon: 140, lead: 200, titanium: 150 },
            size: 3,
            powerConsumption: 4.0,
        }) || this;
    }
    return LaunchPad;
}(Pad));
exports.LaunchPad = LaunchPad;
var LaunchPadLarge = /** @class */ (function (_super) {
    tslib_1.__extends(LaunchPadLarge, _super);
    function LaunchPadLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.requirements = {
            titanium: 200,
            silicon: 150,
            lead: 250,
            plastanium: 75,
        };
        _this.size = 4;
        _this.powerConsumption = 6;
        return _this;
    }
    return LaunchPadLarge;
}(LaunchPad));
exports.LaunchPadLarge = LaunchPadLarge;
var InterplanetaryAccelerator = /** @class */ (function (_super) {
    tslib_1.__extends(InterplanetaryAccelerator, _super);
    function InterplanetaryAccelerator() {
        return _super.call(this, {
            name: 'interplanetary-accelerator',
            requirements: {
                copper: 16000,
                silicon: 11000,
                thorium: 13000,
                titanium: 12000,
                'surge-alloy': 6000,
                'phase-fabric': 5000,
            },
            size: 7,
            powerConsumption: 10.0,
        }) || this;
    }
    InterplanetaryAccelerator.prototype.draw = function (tile, canvas) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var image;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.render({ tile: tile, canvas: canvas, category: category, layers: [this.name] })];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, util_1.blockAsset(category, this.name + '-team')];
                    case 2:
                        image = _a.sent();
                        this.renderImage({
                            canvas: canvas,
                            tile: tile,
                            image: util_1.tintImage(image, '#ffa600'),
                        });
                        return [2 /*return*/];
                }
            });
        });
    };
    return InterplanetaryAccelerator;
}(block_1.Block));
exports.InterplanetaryAccelerator = InterplanetaryAccelerator;
