// mindustry-schematic-parser/dist/mindustry/block/defense.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShockMine = exports.ForceProjector = exports.OverdriveDome = exports.OverdriveProjector = exports.MendProjector = exports.Mender = exports.ScrapWallGigantic = exports.ScrapWallHuge = exports.ScrapWallLarge = exports.ScrapWall = exports.DoorLarge = exports.Door = exports.SurgeWallLarge = exports.SurgeWall = exports.PhaseWallLarge = exports.PhaseWall = exports.ThoriumWallLarge = exports.ThoriumWall = exports.PlastaniumWallLarge = exports.PlastaniumWall = exports.TitaniumWallLarge = exports.TitaniumWall = exports.CopperWallLarge = exports.CopperWall = exports.Wall = void 0;
var tslib_1 = require("tslib");
var block_1 = require("./block");
var category = 'defense';
function multiplyRequirements(requirements, multiplier) {
    if (multiplier === void 0) { multiplier = 4; }
    for (var requirement in requirements) {
        var code = requirement;
        var cost = requirements[code];
        if (cost) {
            requirements[code] = cost * multiplier;
        }
    }
}
var DefenseBlock = /** @class */ (function (_super) {
    tslib_1.__extends(DefenseBlock, _super);
    function DefenseBlock() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DefenseBlock.prototype.draw = function (tile, canvas) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.render({
                            tile: tile,
                            canvas: canvas,
                            category: category,
                            layers: [this.name],
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    return DefenseBlock;
}(block_1.Block));
var Wall = /** @class */ (function (_super) {
    tslib_1.__extends(Wall, _super);
    function Wall(properties) {
        var _this = this;
        var requirements = properties.requirements, size = properties.size;
        _this = _super.call(this, {
            name: properties.name + "-wall",
            requirements: requirements,
            size: size,
        }) || this;
        return _this;
    }
    return Wall;
}(DefenseBlock));
exports.Wall = Wall;
var CopperWall = /** @class */ (function (_super) {
    tslib_1.__extends(CopperWall, _super);
    function CopperWall() {
        return _super.call(this, {
            name: 'copper',
            requirements: {
                copper: 6,
            },
            size: 1,
        }) || this;
    }
    return CopperWall;
}(Wall));
exports.CopperWall = CopperWall;
var CopperWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(CopperWallLarge, _super);
    function CopperWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return CopperWallLarge;
}(CopperWall));
exports.CopperWallLarge = CopperWallLarge;
var TitaniumWall = /** @class */ (function (_super) {
    tslib_1.__extends(TitaniumWall, _super);
    function TitaniumWall() {
        return _super.call(this, {
            name: 'titanium',
            requirements: {
                titanium: 6,
            },
            size: 1,
        }) || this;
    }
    return TitaniumWall;
}(Wall));
exports.TitaniumWall = TitaniumWall;
var TitaniumWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(TitaniumWallLarge, _super);
    function TitaniumWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return TitaniumWallLarge;
}(TitaniumWall));
exports.TitaniumWallLarge = TitaniumWallLarge;
var PlastaniumWall = /** @class */ (function (_super) {
    tslib_1.__extends(PlastaniumWall, _super);
    function PlastaniumWall() {
        return _super.call(this, {
            name: 'plastanium',
            requirements: {
                plastanium: 5,
                metaglass: 2,
            },
            size: 1,
        }) || this;
    }
    return PlastaniumWall;
}(Wall));
exports.PlastaniumWall = PlastaniumWall;
var PlastaniumWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(PlastaniumWallLarge, _super);
    function PlastaniumWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return PlastaniumWallLarge;
}(PlastaniumWall));
exports.PlastaniumWallLarge = PlastaniumWallLarge;
var ThoriumWall = /** @class */ (function (_super) {
    tslib_1.__extends(ThoriumWall, _super);
    function ThoriumWall() {
        return _super.call(this, {
            name: 'thorium',
            requirements: {
                thorium: 6,
            },
            size: 1,
        }) || this;
    }
    return ThoriumWall;
}(Wall));
exports.ThoriumWall = ThoriumWall;
var ThoriumWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(ThoriumWallLarge, _super);
    function ThoriumWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return ThoriumWallLarge;
}(ThoriumWall));
exports.ThoriumWallLarge = ThoriumWallLarge;
var PhaseWall = /** @class */ (function (_super) {
    tslib_1.__extends(PhaseWall, _super);
    function PhaseWall() {
        return _super.call(this, {
            name: 'phase',
            requirements: {
                'phase-fabric': 6,
            },
            size: 1,
        }) || this;
    }
    return PhaseWall;
}(Wall));
exports.PhaseWall = PhaseWall;
var PhaseWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(PhaseWallLarge, _super);
    function PhaseWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return PhaseWallLarge;
}(PhaseWall));
exports.PhaseWallLarge = PhaseWallLarge;
var SurgeWall = /** @class */ (function (_super) {
    tslib_1.__extends(SurgeWall, _super);
    function SurgeWall() {
        return _super.call(this, {
            name: 'surge',
            requirements: {
                'surge-alloy': 6,
            },
            size: 1,
        }) || this;
    }
    return SurgeWall;
}(Wall));
exports.SurgeWall = SurgeWall;
var SurgeWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(SurgeWallLarge, _super);
    function SurgeWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return SurgeWallLarge;
}(SurgeWall));
exports.SurgeWallLarge = SurgeWallLarge;
var Door = /** @class */ (function (_super) {
    tslib_1.__extends(Door, _super);
    function Door() {
        return _super.call(this, {
            name: 'door',
            requirements: {
                titanium: 6,
                silicon: 4,
            },
            size: 1,
        }) || this;
    }
    return Door;
}(DefenseBlock));
exports.Door = Door;
var DoorLarge = /** @class */ (function (_super) {
    tslib_1.__extends(DoorLarge, _super);
    function DoorLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return DoorLarge;
}(Door));
exports.DoorLarge = DoorLarge;
var ScrapWall = /** @class */ (function (_super) {
    tslib_1.__extends(ScrapWall, _super);
    function ScrapWall() {
        return _super.call(this, {
            name: 'scrap',
            requirements: {
                scrap: 6,
            },
            size: 1,
        }) || this;
    }
    return ScrapWall;
}(Wall));
exports.ScrapWall = ScrapWall;
var ScrapWallLarge = /** @class */ (function (_super) {
    tslib_1.__extends(ScrapWallLarge, _super);
    function ScrapWallLarge() {
        var _this = _super.call(this) || this;
        _this.name += '-large';
        _this.size = 2;
        multiplyRequirements(_this.requirements);
        return _this;
    }
    return ScrapWallLarge;
}(ScrapWall));
exports.ScrapWallLarge = ScrapWallLarge;
var ScrapWallHuge = /** @class */ (function (_super) {
    tslib_1.__extends(ScrapWallHuge, _super);
    function ScrapWallHuge() {
        var _this = _super.call(this) || this;
        _this.name += '-huge';
        _this.size = 3;
        multiplyRequirements(_this.requirements, 9);
        return _this;
    }
    return ScrapWallHuge;
}(ScrapWall));
exports.ScrapWallHuge = ScrapWallHuge;
var ScrapWallGigantic = /** @class */ (function (_super) {
    tslib_1.__extends(ScrapWallGigantic, _super);
    function ScrapWallGigantic() {
        var _this = _super.call(this) || this;
        _this.name += '-gigantic';
        _this.size = 4;
        multiplyRequirements(_this.requirements, 16);
        return _this;
    }
    return ScrapWallGigantic;
}(ScrapWall));
exports.ScrapWallGigantic = ScrapWallGigantic;
var Mender = /** @class */ (function (_super) {
    tslib_1.__extends(Mender, _super);
    function Mender() {
        return _super.call(this, {
            name: 'mender',
            requirements: { lead: 30, copper: 25 },
            size: 1,
            powerConsumption: 0.3,
        }) || this;
    }
    return Mender;
}(DefenseBlock));
exports.Mender = Mender;
var MendProjector = /** @class */ (function (_super) {
    tslib_1.__extends(MendProjector, _super);
    function MendProjector() {
        return _super.call(this, {
            name: 'mend-projector',
            requirements: { lead: 100, titanium: 25, silicon: 40 },
            size: 2,
            powerConsumption: 1.5,
        }) || this;
    }
    return MendProjector;
}(DefenseBlock));
exports.MendProjector = MendProjector;
var OverdriveProjector = /** @class */ (function (_super) {
    tslib_1.__extends(OverdriveProjector, _super);
    function OverdriveProjector() {
        return _super.call(this, {
            name: 'overdrive-projector',
            requirements: { lead: 100, titanium: 75, silicon: 75, plastanium: 30 },
            size: 2,
            powerConsumption: 3.5,
        }) || this;
    }
    return OverdriveProjector;
}(DefenseBlock));
exports.OverdriveProjector = OverdriveProjector;
var OverdriveDome = /** @class */ (function (_super) {
    tslib_1.__extends(OverdriveDome, _super);
    function OverdriveDome() {
        return _super.call(this, {
            name: 'overdrive-dome',
            requirements: {
                lead: 200,
                titanium: 130,
                silicon: 130,
                plastanium: 80,
                'surge-alloy': 120,
            },
            size: 3,
            powerConsumption: 10.0,
        }) || this;
    }
    return OverdriveDome;
}(DefenseBlock));
exports.OverdriveDome = OverdriveDome;
var ForceProjector = /** @class */ (function (_super) {
    tslib_1.__extends(ForceProjector, _super);
    function ForceProjector() {
        return _super.call(this, {
            name: 'force-projector',
            requirements: { lead: 100, titanium: 75, silicon: 125 },
            size: 3,
            powerConsumption: 4.0,
        }) || this;
    }
    return ForceProjector;
}(DefenseBlock));
exports.ForceProjector = ForceProjector;
var ShockMine = /** @class */ (function (_super) {
    tslib_1.__extends(ShockMine, _super);
    function ShockMine() {
        return _super.call(this, {
            name: 'shock-mine',
            requirements: { lead: 25, silicon: 12 },
            size: 1,
        }) || this;
    }
    return ShockMine;
}(DefenseBlock));
exports.ShockMine = ShockMine;
