// mindustry-schematic-parser/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Schematic = exports.SchematicTile = exports.Liquids = exports.Items = exports.Blocks = exports.mindustry = exports.arc = void 0;
var tslib_1 = require("tslib");
exports.arc = tslib_1.__importStar(require("./arc"));
exports.mindustry = tslib_1.__importStar(require("./mindustry"));
var mindustry_1 = require("./mindustry");
Object.defineProperty(exports, "Blocks", { enumerable: true, get: function () { return mindustry_1.Blocks; } });
Object.defineProperty(exports, "Items", { enumerable: true, get: function () { return mindustry_1.Items; } });
Object.defineProperty(exports, "Liquids", { enumerable: true, get: function () { return mindustry_1.Liquids; } });
var schematic_1 = require("./schematic");
Object.defineProperty(exports, "SchematicTile", { enumerable: true, get: function () { return schematic_1.SchematicTile; } });
Object.defineProperty(exports, "Schematic", { enumerable: true, get: function () { return schematic_1.Schematic; } });
