// mindustry-schematic-parser/dist/mindustry/content.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnlockableContent = exports.MappableContent = exports.Content = void 0;
var tslib_1 = require("tslib");
/** An abstract top class representing in game content
 *
 * This class should not be instantiated
 */
var Content = /** @class */ (function () {
    function Content() {
    }
    return Content;
}());
exports.Content = Content;
/** An abstract top class representing idetifiable in game content
 *
 * This class should not be instantiated
 */
var MappableContent = /** @class */ (function (_super) {
    tslib_1.__extends(MappableContent, _super);
    function MappableContent(name) {
        var _this = _super.call(this) || this;
        _this.name = name;
        return _this;
    }
    MappableContent.prototype.toString = function () {
        return this.name;
    };
    return MappableContent;
}(Content));
exports.MappableContent = MappableContent;
/** An abstract top class representing unlockable in game content
 *
 * This class should not be instantiated
 */
var UnlockableContent = /** @class */ (function (_super) {
    tslib_1.__extends(UnlockableContent, _super);
    function UnlockableContent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return UnlockableContent;
}(MappableContent));
exports.UnlockableContent = UnlockableContent;
