// mindustry-schematic-parser/dist/mindustry/liquid/liquid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Liquid = void 0;
var liquidColors = {
    water: '#596ab8',
    slag: '#ffa166',
    oil: '#313131',
    cryofluid: '#6ecdec',
};
var Liquid = /** @class */ (function () {
    function Liquid(name) {
        this.name = name;
        this.color = liquidColors[name];
    }
    Liquid.create = function (name) {
        var liquid = this.liquidMap.get(name);
        if (!liquid) {
            liquid = new Liquid(name);
            this.liquidMap.set(name, liquid);
        }
        return liquid;
    };
    Liquid.fromCode = function (code) {
        var _a;
        var liquids = ['water', 'slag', 'oil', 'cryofluid'];
        if (code > liquids.length - 1)
            throw new Error('Unknown liquid code: ' + code);
        var name = liquids[code];
        return (_a = this.liquidMap.get(name)) !== null && _a !== void 0 ? _a : this.create(name);
    };
    Liquid.liquidMap = new Map();
    return Liquid;
}());
exports.Liquid = Liquid;
